### DISCUSSION TITLE
Memory use by daemons

### DISCUSSION THREAD

**Discussion by hapatrick** [nolar/kopf#951] [Q&A] [answered]
We have a daemon that handles a custom resource called Workflow. Workflow objects manage batch processing. Once the processing associated with a Workflow finishes, the daemon exits. But the Workflow object stays in the api server so it can be examined later for the processing status / results.

I know that daemons in kopf are *intended* to run for the lifetime of an object, but you can return from the daemon handler and I assumed that would destroy any resources needed for the daemon.

But what we are seeing is that our operator's memory use keeps going up as we create new Workflow objects -- and goes down when we delete Workflow objects from the api server. Which seems to imply that kopf is using some memory for each Workflow object, even the ones whose daemon handlers have exited.

Is it expected that kopf would use memory for each object in the api server that has an associated daemon handler, even if the daemon handler has exited?

**Comment by nolar** [comment, marked as answer]
Hi. Thanks for asking.

Yes, Kopf keeps track of which daemons have actually exited — to not start them again. If that list is deleted, the daemons will be respawned as for a new object. But that memory should be minimal, a few bytes. Here is the list of daemon/handler ids per resource: https://github.com/nolar/kopf/blob/1.35.6/kopf/_core/engines/daemons.py#L416-L429

What could be leaking the memory, is the "live body" object of the resource. It is remembered once the daemon is started, updated on every new event from Kubernetes: https://github.com/nolar/kopf/blob/7b4569024b5a9382195bd0ba76ffb164c2c41bbd/kopf/_core/reactor/processing.py#L54-L65

I think, there is no place where `.live_fresh_body` is set to None. It is a good idea to add it somewhere around the 1st place mentioned above (the `finally:` block) — but only of ALL daemons of that resource X have stopped. I think, checking `memory.running_daemons` (minus the current daemon) should be a sufficient criterion for this.

Can you please make a PR?

**Reply by hapatrick** [reply]
Thanks. We tried this out and it did indeed work. One of my colleagues is going to post a PR with the change.

**Reply by hapatrick** [reply]
PR here: https://github.com/nolar/kopf/pull/953
